wi-fi/lwip/main.c

```diff
@@ -0,0 +1,293 @@
+	if (result != CY_RSLT_SUCCESS) {
+		wm_cy_log_msg(CYLF_MIDDLEWARE, CY_LOG_ERR, "can't start Wi-Fi AP\n");
+		cy_lwip_network_down(&wifi_common.iface);
+		cy_lwip_remove_interface(&wifi_common.iface);
+		cybsp_wifi_deinit(wifi_common.iface.whd_iface);
+		cybsp_free();
+		return -6;
+	}
+
+	wifi_common.running = true;
+
+	return 0;
+}
+
+
+void wifi_ap_stop(void)
+{
+	if (!wifi_common.running) {
+		wm_cy_log_msg(CYLF_MIDDLEWARE, CY_LOG_WARNING, "Wi-Fi is already stopped\n");
+		return;
+	}
+
+	whd_wifi_stop_ap(wifi_common.iface.whd_iface);
+	cy_lwip_network_down(&wifi_common.iface);
+	cy_lwip_remove_interface(&wifi_common.iface);
+	cybsp_wifi_deinit(wifi_common.iface.whd_iface);
+	cybsp_free();
+
+	wifi_common.running = false;
+}
+
+
+static int wifi_dev_open(int flags)
+{
+	char *buf;
+	size_t size;
+	int overflow = 0;
+
+	if (wifi_common.dev.busy)
+		return -EBUSY;
+
+	buf = wifi_common.dev.buf;
+	size = sizeof(wifi_common.dev.buf);
+
+	SNPRINTF_APPEND("running=%u\n", wifi_common.running);
+
+	if (overflow)
+		return -EFBIG;
+
+	wifi_common.dev.busy = 1;
+	wifi_common.dev.len = buf - wifi_common.dev.buf;
+
+	return 0;
+}
+
+
+static int wifi_dev_close(void)
+{
+	if (!wifi_common.dev.busy)
+		return -EBADF;
+	wifi_common.dev.busy = 0;
+	return 0;
+}
+
+
+static int wifi_dev_read(char *data, size_t size, size_t offset)
+{
+	int read;
+
+	if (offset > wifi_common.dev.len)
+		return -ERANGE;
+
+	read = min(size, wifi_common.dev.len - offset);
+	memcpy(data, wifi_common.dev.buf + offset, read);
+
+	return read;
+}
+
+
+static int wifi_dev_write(char *data, size_t size)
+{
+	if (strncmp("start", data, size) == 0) {
+		wifi_ap_start();
+	}
+	else if (strncmp("stop", data, size) == 0) {
+		wifi_ap_stop();
+	}
+	else {
+		wm_cy_log_msg(CYLF_MIDDLEWARE, CY_LOG_ERR, "got unknown Wi-Fi command\n");
+		return -EINVAL;
+	}
+
+	return size;
+}
+
+
+int wifi_dev_init(unsigned int port)
+{
+	oid_t wifi_oid = { port, WIFI_DEV_ID };
+
+	return create_dev(&wifi_oid, WIFI_DEV_NAME);
+}
+
+
+static void wifi_thread(void *arg)
+{
+	cy_rslt_t result;
+	unsigned int port;
+
+	result = cy_log_init(CY_LOG_DEBUG);
+	if (result != CY_RSLT_SUCCESS) {
+		printf("phoenix-rtos-lwip: can't init Wi-Fi logs\n");
+		return;
+	}
+
+	if (portCreate(&port) < 0) {
+		wm_cy_log_msg(CYLF_MIDDLEWARE, CY_LOG_ERR, "can't create Wi-Fi port\n");
+		return;
+	}
+
+	if (wifi_dev_init(port) < 0) {
+		wm_cy_log_msg(CYLF_MIDDLEWARE, CY_LOG_ERR, "can't create Wi-Fi device\n");
+		return;
+	}
+
+	for (;;) {
+		msg_t msg = { 0 };
+		unsigned long int rid;
+
+		if (msgRecv(port, &msg, &rid) < 0)
+			continue;
+
+		switch (msg.type) {
+			case mtOpen:
+				msg.o.io.err = wifi_dev_open(msg.i.openclose.flags);
+				break;
+
+			case mtClose:
+				msg.o.io.err = wifi_dev_close();
+				break;
+
+			case mtRead:
+				msg.o.io.err = wifi_dev_read(msg.o.data, msg.o.size, msg.i.io.offs);
+				break;
+
+			case mtWrite:
+				msg.o.io.err = wifi_dev_write(msg.i.data, msg.i.size);
+				break;
+
+			default:
```

Comment:
It would be safer to return `-EINVAL` in the `msg.i.io.err` if we don't handle a particular message type.